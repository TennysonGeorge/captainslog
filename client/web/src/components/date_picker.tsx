import * as React from "react"
import ReactDatePicker from "react-datepicker"

/* tslint:disable:no-var-requires */
require("../react-datepicker.css")
/* tslint:enable:no-var-requires */

// This should always mimic the Grouping type found in model/book.go.
export enum Grouping {
  NONE,
  DAY,
}

enum Unit {
  NONE = 0,
  DAY = 1000 * 60 * 60 * 24,
}

const GroupUnit: { [index: number]: Unit } = {
  [Grouping.NONE]: Unit.NONE,
  [Grouping.DAY]: Unit.DAY,
}

const add = (date: Date, unit: Unit): Date =>
  new Date(+date + unit)

const sub = (date: Date, unit: Unit): Date =>
  new Date(+date - unit)

const Btn = ({ label, action }: { label: string, action: () => void }) =>
  <input type="button" onClick={action} value={label} />

type DatePickerProps = {
  date: Date
  onChange: (_: Date) => void
  grouping: Grouping
}

export const DatePicker = ({ grouping, date, onChange }: DatePickerProps) => {
  const unit = GroupUnit[grouping] || Unit.NONE
  const handler = (maybeDate: Date | null) =>
    maybeDate ? onChange(maybeDate) : null

  return <div>
    <ReactDatePicker selected={date} onChange={handler} />
    <Btn label="prev" action={() => onChange(sub(date, unit))} />
    <Btn label="next" action={() => onChange(add(date, unit))} />
    <Btn label="current" action={() => onChange(new Date())} />
  </div>
}
