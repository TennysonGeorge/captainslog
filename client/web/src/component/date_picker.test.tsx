import * as React from "react"
import { act } from "react-dom/test-utils"
import { mount } from "enzyme"

import "./testsetup"

import { DatePicker, Grouping } from "./date_picker"

test("snapshot DatePicker (no grouping)", () => {
  const date = new Date(1566096667888)
  const onChange = (_: Date) => {}
  const grouping = Grouping.NONE
  const component = mount(<DatePicker date={date} onChange={onChange} grouping={grouping} />)
  expect(component).toMatchSnapshot()
})

test("snapshot DatePicker (daily grouping)", () => {
  const date = new Date(1566096667888)
  const onChange = (_: Date) => {}
  const grouping = Grouping.DAY
  const component = mount(<DatePicker date={date} onChange={onChange} grouping={grouping} />)
  expect(component).toMatchSnapshot()
})
