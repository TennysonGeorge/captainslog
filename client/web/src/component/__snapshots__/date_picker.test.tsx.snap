// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker (daily grouping) 1`] = `
<Component
  date={2019-08-18T02:51:07.888Z}
  grouping={1}
  onChange={[Function]}
>
  <DatePicker
    allowSameDay={false}
    dateFormat="MM/dd/yyyy"
    dateFormatCalendar="LLLL yyyy"
    disabled={false}
    disabledKeyboardNavigation={false}
    dropdownMode="scroll"
    inlineFocusSelectedMonth={false}
    monthsShown={1}
    nextMonthButtonLabel="Next month"
    onBlur={[Function]}
    onChange={[Function]}
    onClickOutside={[Function]}
    onFocus={[Function]}
    onInputClick={[Function]}
    onInputError={[Function]}
    onKeyDown={[Function]}
    onMonthChange={[Function]}
    onSelect={[Function]}
    onYearChange={[Function]}
    preventOpenOnFocus={false}
    previousMonthButtonLabel="Previous Month"
    readOnly={false}
    renderDayContents={[Function]}
    selected={2019-08-18T02:51:07.888Z}
    shouldCloseOnSelect={true}
    showMonthYearPicker={false}
    showTimeInput={false}
    showTimeSelect={false}
    strictParsing={false}
    timeCaption="Time"
    timeInputLabel="Time"
    timeIntervals={30}
    withPortal={false}
  >
    <PopperComponent
      hidePopper={true}
      popperComponent={null}
      popperModifiers={
        Object {
          "preventOverflow": Object {
            "boundariesElement": "viewport",
            "enabled": true,
            "escapeWithReference": true,
          },
        }
      }
      popperPlacement="bottom-start"
      popperProps={Object {}}
      targetComponent={
        <div
          className="react-datepicker__input-container"
        >
          <input
            className=""
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            readOnly={false}
            type="text"
            value="08/17/2019"
          />
        </div>
      }
    >
      <Manager>
        <Reference>
          <InnerReference
            setReferenceNode={[Function]}
          >
            <div
              className="react-datepicker-wrapper"
            >
              <div
                className="react-datepicker__input-container"
              >
                <input
                  className=""
                  disabled={false}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  readOnly={false}
                  type="text"
                  value="08/17/2019"
                />
              </div>
            </div>
          </InnerReference>
        </Reference>
      </Manager>
    </PopperComponent>
  </DatePicker>
  <Btn
    action={[Function]}
    label="Previous"
  >
    <input
      onClick={[Function]}
      type="button"
      value="Previous"
    />
  </Btn>
  <Btn
    action={[Function]}
    label="Next"
  >
    <input
      onClick={[Function]}
      type="button"
      value="Next"
    />
  </Btn>
  <Btn
    action={[Function]}
    label="Current"
  >
    <input
      onClick={[Function]}
      type="button"
      value="Current"
    />
  </Btn>
</Component>
`;

exports[`DatePicker (no grouping) 1`] = `
<Component
  date={2019-08-18T02:51:07.888Z}
  grouping={0}
  onChange={[Function]}
>
  <DatePicker
    allowSameDay={false}
    dateFormat="MM/dd/yyyy"
    dateFormatCalendar="LLLL yyyy"
    disabled={false}
    disabledKeyboardNavigation={false}
    dropdownMode="scroll"
    inlineFocusSelectedMonth={false}
    monthsShown={1}
    nextMonthButtonLabel="Next month"
    onBlur={[Function]}
    onChange={[Function]}
    onClickOutside={[Function]}
    onFocus={[Function]}
    onInputClick={[Function]}
    onInputError={[Function]}
    onKeyDown={[Function]}
    onMonthChange={[Function]}
    onSelect={[Function]}
    onYearChange={[Function]}
    preventOpenOnFocus={false}
    previousMonthButtonLabel="Previous Month"
    readOnly={false}
    renderDayContents={[Function]}
    selected={2019-08-18T02:51:07.888Z}
    shouldCloseOnSelect={true}
    showMonthYearPicker={false}
    showTimeInput={false}
    showTimeSelect={false}
    strictParsing={false}
    timeCaption="Time"
    timeInputLabel="Time"
    timeIntervals={30}
    withPortal={false}
  >
    <PopperComponent
      hidePopper={true}
      popperComponent={null}
      popperModifiers={
        Object {
          "preventOverflow": Object {
            "boundariesElement": "viewport",
            "enabled": true,
            "escapeWithReference": true,
          },
        }
      }
      popperPlacement="bottom-start"
      popperProps={Object {}}
      targetComponent={
        <div
          className="react-datepicker__input-container"
        >
          <input
            className=""
            disabled={false}
            onBlur={[Function]}
            onChange={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            readOnly={false}
            type="text"
            value="08/17/2019"
          />
        </div>
      }
    >
      <Manager>
        <Reference>
          <InnerReference
            setReferenceNode={[Function]}
          >
            <div
              className="react-datepicker-wrapper"
            >
              <div
                className="react-datepicker__input-container"
              >
                <input
                  className=""
                  disabled={false}
                  onBlur={[Function]}
                  onChange={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  readOnly={false}
                  type="text"
                  value="08/17/2019"
                />
              </div>
            </div>
          </InnerReference>
        </Reference>
      </Manager>
    </PopperComponent>
  </DatePicker>
  <Btn
    action={[Function]}
    label="Previous"
  >
    <input
      onClick={[Function]}
      type="button"
      value="Previous"
    />
  </Btn>
  <Btn
    action={[Function]}
    label="Next"
  >
    <input
      onClick={[Function]}
      type="button"
      value="Next"
    />
  </Btn>
  <Btn
    action={[Function]}
    label="Current"
  >
    <input
      onClick={[Function]}
      type="button"
      value="Current"
    />
  </Btn>
</Component>
`;
